test/functional/bind.cpp

```diff
@@ -0,0 +1,387 @@
+  CHECK(std::is_base_of<slb::integral_constant<int, 0>,
+                        slb::is_placeholder<int>>::value);
+}
+
+// template<class F, class... BoundArgs>
+//   unspecified bind(F&&, BoundArgs&&...);
+// template<class R, class F, class... BoundArgs>
+//   unspecified bind(F&&, BoundArgs&&...);
+struct C {
+  int obj;
+  C(int val) : obj(val) {}
+
+  int fun(int base) noexcept { return base + 0; }
+  int cfun(int base) const noexcept { return base + 1; }
+  int lfun(int base) & noexcept { return base + 2; }
+  int rfun(int base) && noexcept { return base + 3; }
+  int clfun(int base) const& noexcept { return base + 4; }
+  int crfun(int base) const&& noexcept { return base + 5; }
+};
+
+struct D : C {
+  D(int val) : C(val) {}
+};
+
+template <typename T>
+struct smart_ptr {
+  T* ptr;
+  smart_ptr(T* ptr) : ptr(ptr) {}
+  T& operator*() const noexcept { return *ptr; }
+};
+
+template <typename R,
+          typename Bound,
+          typename BoundR,
+          typename BoundVoid,
+          typename... Args>
+void check_bound(
+    R&& r, Bound bound, BoundR bound_r, BoundVoid bound_void, Args&&... args) {
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound)>>::value);
+  CHECK(bound(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_r)>>::value);
+  CHECK(std::is_same<decltype(bound_r(std::forward<Args>(args)...)), R>::value);
+  CHECK(bound_r(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_void)>>::value);
+  CHECK(std::is_same<decltype(bound_void(std::forward<Args>(args)...)),
+                     void>::value);
+  bound_void(std::forward<Args>(args)...);
+}
+
+template <typename R,
+          typename Bound,
+          typename BoundR,
+          typename BoundVoid,
+          typename... Args>
+void check_bound_const(R&& r,
+                       Bound const bound,
+                       BoundR const bound_r,
+                       BoundVoid const bound_void,
+                       Args&&... args) {
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound)>>::value);
+  CHECK(bound(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_r)>>::value);
+  CHECK(std::is_same<decltype(bound_r(std::forward<Args>(args)...)), R>::value);
+  CHECK(bound_r(std::forward<Args>(args)...) == r);
+
+  CHECK(std::is_base_of<slb::true_type,
+                        slb::is_bind_expression<decltype(bound_void)>>::value);
+  CHECK(std::is_same<decltype(bound_void(std::forward<Args>(args)...)),
+                     void>::value);
+  bound_void(std::forward<Args>(args)...);
+}
+
+template <typename R, typename F, typename Arg1>
+void check_bind(R&& r, F&& f, Arg1&& arg1) {
```

Comment:
Can we break this into `check_bind_obj` and `check_bind_fun` similar to what we did for `invoke`?
This way we don't have to deal with the oddity of "sometimes specified `R`".